diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/oom-listener/impl/oom_listener.c b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/oom-listener/impl/oom_listener.c
index 0086b26cd47..ece270f27c8 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/oom-listener/impl/oom_listener.c
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/oom-listener/impl/oom_listener.c
@@ -22,6 +22,19 @@
 #include <poll.h>
 #include "oom_listener.h"
 
+/*
+ Clean up allocated resources in a descriptor structure
+*/
+void cleanup(_oom_listener_descriptors *descriptors) {
+  close(descriptors->event_fd);
+  descriptors->event_fd = -1;
+  close(descriptors->event_control_fd);
+  descriptors->event_control_fd = -1;
+  close(descriptors->oom_control_fd);
+  descriptors->oom_control_fd = -1;
+  descriptors->watch_timeout = 1000;
+}
+
 /*
  * Print an error.
 */
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/oom-listener/impl/oom_listener.h b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/oom-listener/impl/oom_listener.h
index aa77cb66f58..b11f4f1e167 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/oom-listener/impl/oom_listener.h
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/oom-listener/impl/oom_listener.h
@@ -81,15 +81,7 @@ typedef struct _oom_listener_descriptors {
 /*
  Clean up allocated resources in a descriptor structure
 */
-inline void cleanup(_oom_listener_descriptors *descriptors) {
-  close(descriptors->event_fd);
-  descriptors->event_fd = -1;
-  close(descriptors->event_control_fd);
-  descriptors->event_control_fd = -1;
-  close(descriptors->oom_control_fd);
-  descriptors->oom_control_fd = -1;
-  descriptors->watch_timeout = 1000;
-}
+void cleanup(_oom_listener_descriptors *descriptors);
 
 /*
  * Enable an OOM listener on the memory cgroup cgroup
